### ISSUE TITLE
TryAsync for exception-throwing Tasks

### ISSUE THREAD

**Issue by NickDarvey** [louthy/language-ext#201]
Hi Paul,

Would you be open to an addition or extension related to Try that handled Tasks?

The current implementation (1.9.7) won't catch exceptions because it's not awaited (and async methods return immediately, etc. etc.) and it looks like it's similar in 2.0.0.

Is there a better way to be handling these cases? If not and it's something you'd like to include, do you have a preference for (roughly) where I ought to add this if I were to do this as a PR?

Nick

**Comment by louthy** [comment]
@NickDarvey Funnily enough I was thinking the other day about how successful the `CondAsync` and `Cond` pairing is.  The asynchronous aspects of it are transparent, and it would be great to have `OptionAsync`, `TryAsync`, etc.  

It's certainly not free though - in terms of implementation time and long term maintenance.  But it's certainly a better approach than current extensions `Task<Try<A>>`, `Try<Task<A>>`

Do you have a specific repro I could look at?

**Comment by NickDarvey** [comment]
I haven't created anything yet (other than something we slapped together for work) but I'll have a shot this weekend at a decent implementation and let you know how I go.

Where is the `Try<Task<A>>` extension btw?

**Comment by louthy** [comment]
@NickDarvey Hi Nick, I have a solution.  So currently `Try<A>` has a few `MatchAsync` methods that allow you to do various combinations of asynchronous calls.  There are extensions for `Task<Try<A>>` and variants that allow the `Succ` and `Fail` branches to return `Task<B>`.  That's obviously the minimal behaviour you'd want to get a value out of a `Try` asynchronously.

The ideal solution would be:

* Every `Try` operation has an `Async` variant
* A `TryAsync<A>` which provides asynchronous behaviour by default
* Extensions to `Task<Try<A>>` that make it work exactly like `TryAsync<A>`

So that's what I've done, and is available as `v2.0.19-alpha` on nuget now.

For example:
```c#
        // Some example method prototypes
        public TryAsync<int> LongRunningAsyncTaskOp() => TryAsync(10);
        public Task<Try<int>> LongRunningAsyncOp()    => Task.FromResult(Try(10));
        public Try<int> LongRunningOp()               => () => 10;
        public Task<int> MapToTask(int x)             => Task.FromResult(x * 2);
        public int MapTo(int x)                       => x * 2;

        public async void Test()
        {
            // These run synchronously
            int a = LongRunningOp().IfFail(0);
            Try<int> b = LongRunningOp().Map(MapTo);

            // These run asynchronously
            int u = await LongRunningAsyncTaskOp().IfFail(0);
            int v = await LongRunningAsyncOp().IfFail(0);
            int x = await LongRunningOp().IfFailAsync(0);
            TryAsync<int> y1 = LongRunningOp().MapAsync(MapTo);
            TryAsync<int> y2 = LongRunningOp().MapAsync(MapToTask);
            int z1 = await y1.IfFail(0);
            int z2 = await y2.IfFail(0);
        }
```
As much as possible I've tried to make the step from `Try<A>` to `TryAsync<A>` transparent.  So if you return a `Task<R>` in the `Map` function of a `Try`, then it gets 'upgraded' to a `TryAsync`.  This is similar behaviour to the `Cond` type.

Also where tasks can run in parallel they will.  So for example, if you use the new applicatives:
```c#
        TryAsync<int> GetValueOne() => TryAsync(2);
        TryAsync<int> GetValueTwo() => TryAsync(() => 10);

        public async void ApplicativeTest()
        {
            var op = apply(
                TryAsync<Func<int, int, int>>((x, y) => x + y),
                GetValueOne(),
                GetValueTwo()
            );

            var res = await op.IfFail(0);
            Assert.True(res == 12);
        }
```
Obviously that's a basic example, but `GetValueOne` and `GetValueTwo` will run in parallel, check failure states automatically, and then use the applicative functor to combine the values.  You could imagine the two functions doing something heavyweight that returned lists of results, which are then reduced.  And for the avoidance of doubt the 'reduction' would be asynchronous too.

One final point is that you can make a `Try<A>` into a `TryAsync<A>` at any point by calling:
```c#
     TryAsync<A> x = tryValue.ToAsync();
```
And you can convert a `Task<Try<A>>` or a `Try<Task<A>>` to a `TryAsync<A>` by:
```c#
     TryAsync<A> x = tryTask.ToTryAsync();
```
Again it will deal with all of the possible failures.  

I think that should cover all possible combinations of `Try` and `Task` and provide convenience functions for when the type isn't in the state you want.  I still need to do this for `TryOption` (it's a lot of typing!), I'll try to get to that within the next week or so.

You will need to be on version 2, which although is alpha, is stable.  But if you're using the `Process` system then you should probably hold off until I've done some more testing.

**Comment by NickDarvey** [comment]
>So that's what I've done, and is available as v2.0.19-alpha on nuget now.
Bloody hell you are fast. This is fantastic!

**Comment by louthy** [comment]
@NickDarvey `v2.0.20-alpha` is up with `TryOptionAsync`, and all the same extensions as listed above for `Try`.  

A couple of things to note:

1. `ToTryAsync` has been renamed to `ToAsync`
2. `TryOption.IfNone(...)` (which actually handled `None` and `Fail` states), has now been renamed to `IfNoneOrFail`.  the naming was confusing I think, so it's explicit now.  Obviously that will be a breaking change if you use `TryOption` at all.

Here's the same version of the examples above for `TryAsync` and `TryOptionAsync`.  

```c#
class TryAsyncExamples
{ 
    // Some example method prototypes
    public TryAsync<int> LongRunningAsyncTaskOp() => TryAsync(10);
    public Task<Try<int>> LongRunningAsyncOp()    => Task.FromResult(Try(10));
    public Try<int> LongRunningOp()               => Try(10);
    public Task<int> MapToTask(int x)             => Task.FromResult(x * 2);
    public int MapTo(int x)                       => x * 2;

    public async void Test()
    {
        TryAsync<int> j = LongRunningOp().ToAsync();
        TryAsync<int> k = LongRunningAsyncOp().ToAsync();

        // These run synchronously
        int a = LongRunningOp().IfFail(0);
        Try<int> b = LongRunningOp().Map(MapTo);

        // These run asynchronously
        int u = await LongRunningAsyncTaskOp().IfFail(0);
        int v = await LongRunningAsyncOp().IfFail(0);
        int x = await LongRunningOp().IfFailAsync(0);
        TryAsync<int> y1 = LongRunningOp().MapAsync(MapTo);
        TryAsync<int> y2 = LongRunningOp().MapAsync(MapToTask);
        int z1 = await y1.IfFail(0);
        int z2 = await y2.IfFail(0);
    }
}

class TryOptionAsyncExamples
{
    // Some example method prototypes
    public TryOptionAsync<int> LongRunningAsyncTaskOp() => TryOptionAsync(10);
    public Task<TryOption<int>> LongRunningAsyncOp()    => Task.FromResult(TryOption(10));
    public TryOption<int> LongRunningOp()               => TryOption(10);
    public Task<int> MapToTask(int x)                   => Task.FromResult(x * 2);
    public int MapTo(int x)                             => x * 2;

    public async void Test()
    {
        TryOptionAsync<int> j = LongRunningOp().ToAsync();
        TryOptionAsync<int> k = LongRunningAsyncOp().ToAsync();

        // These run synchronously
        int a = LongRunningOp().IfNoneOrFail(0);
        TryOption<int> b = LongRunningOp().Map(MapTo);

        // These run asynchronously
        int u = await LongRunningAsyncTaskOp().IfNoneOrFail(0);
        int v = await LongRunningAsyncOp().IfNoneOrFail(0);
        int x = await LongRunningOp().IfNoneOrFailAsync(0);
        TryOptionAsync<int> y1 = LongRunningOp().MapAsync(MapTo);
        TryOptionAsync<int> y2 = LongRunningOp().MapAsync(MapToTask);
        int z1 = await y1.IfNoneOrFail(0);
        int z2 = await y2.IfNoneOrFail(0);
    }
}
```
Please let me know how you get on, and if you have any problems.  I rarely use `Task` myself (because I use the lang-ext Process system).

**Comment by ticofab** [comment]
Hi @louthy , Thanks for your work, I come from a JVM background and this library is helping. At the same time, there are some things that I can't get to work. For instance, does this example still work? I cannot make anything of this compile. In particular, I would like to achieve something like
```
Task<string> task = _httpClient.GetStringAsync(url);
---> how do I create a Try / TryAsync around this Task?
```

Basically I am looking for the same answer as this question on gitter:
https://gitter.im/louthy/language-ext/archives/2017/06/29

Thanks for any help!

**Comment by bender2k14** [comment]
Can you post your question as a new issue?
